File: Data Structures and Algorithms/Algorithms/c/heapsort.c
#include<stdio.h>

void swap(int *xp, int *yp) 
{ 
    int temp = *xp; 
    *xp = *yp; 
    *yp = temp; 
}

void heapify(int ar[], int n, int i)
{
    int largest = i;
    int l = 2*i+1;
    int r = 2*i+2;

    if(l<n&&ar[l]>ar[largest])
        largest = l;

    if(r<n&&ar[r]>ar[largest])
        largest = r;
    
    if(largest!=i)  
    {
        swap(&ar[i],&ar[largest]); heapify(ar,n,largest);
    }
}

void heapSort(int ar[], int n)
{
    for(int i=n/2-1; i>=0; i--)
        heapify(ar,n,i);
    
    for(int i=n-1; i>0; i--)
    {
        swap(&ar[0],&ar[i]);
        
        heapify(ar,i,0);
    }
}

void printArray(int ar[], int n)
{
    for(int i=0; i<n; i++)
        printf("%d ",ar[i]);
    
    printf("\n");
}

int main()
{
    int ar[]={12,11,13,5,6,7};
    int n=sizeof(ar)/sizeof(ar[0]);

    heapSort(ar,n);

    printf("Sorted Array is: \n");
    
    printArray(ar,n);
    return 0;
}

File: Python/heap_sort.py
def heapify(arr, n, i): 
    largest = i
    l = 2 * i + 1
    r = 2 * i + 2

    if l < n and arr[i] < arr[l]:
      largest = l
    
    if r < n and arr[largest] < arr[r]:
      largest = r


    if largest != i:
      arr[i], arr[largest] = arr[largest], arr[i]
      heapify(arr, n, largest)

def heapSort(arr):
  n = len(arr)

  for i in range(n //2, -1, -1):
    heapify(arr, n, i)

  for i in range(n - 1, 0, -1): 
    arr[i], arr[0] = arr[0], arr[i]
    heapify(arr, i, 0)

arr = [11, 34, 9, 5, 16, 10] 
n = len(arr) 
print("Original array:") 
for i in range(n):       
    print("%d " % arr[i], end = '')
heapSort(arr) 
print("Sorted array:") 
for i in range(n):       
    print("%d " % arr[i], end = '')
